File: Algorithms/Python/0022_Generate_Parentheses/Generate_Parentheses_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def generateParenthesis(self, n):
        if n == 0: return ['']
        if n == 1: return ['()']
        self.res = []

        def backtracking(n, left_count, right_count, temp_res):
            if right_count > left_count: return
            if left_count == right_count == n:
                self.res.append(temp_res)
                return

            for each in '()':
                if each == '(' and left_count < n:
                    left_count += 1
                    temp_res += '('
                    backtracking(n, left_count, right_count, temp_res)
                    temp_res = temp_res[:-1]
                    left_count -= 1
                if each == ')' and right_count < n:
                    right_count += 1
                    temp_res += ')'
                    backtracking(n, left_count, right_count, temp_res)
                    temp_res = temp_res[:-1]
                    right_count -= 1

        backtracking(n, 0, 0, '')
        return self.res

File: Algorithms/0022_Generate_Parentheses/Java/GenerateParenthesesSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    private Set<String> res = new HashSet<>();

    public List<String> generateParenthesis(int n) {
        if (n == 0) return new ArrayList<>();
        if (n == 1) return Arrays.asList("()");

        dfs(n, n, "");
        return new ArrayList<>(res);
    }

    private void dfs(int left, int right, String tempRes) {
        if (left == right && left == 0) {
            if (verify(tempRes)) {
                res.add(tempRes);
            }
            return;
        }
        if (left > 0 && left <= right) {
            tempRes += "(";
            dfs(left - 1, right, tempRes);
            tempRes = tempRes.substring(0, tempRes.length() - 1);
            tempRes += ")";
            dfs(left, right - 1, tempRes);
            tempRes = tempRes.substring(0, tempRes.length() - 1);
        }
        if (left == 0) {
            tempRes += ")";
            dfs(left, right - 1, tempRes);
            tempRes = tempRes.substring(0, tempRes.length() - 1);
        }
    }

    private boolean verify(String str) {
        int count = 0;
        for (char c : str.toCharArray()) {
            if (c == '(') count++;
            else if (c == ')') count--;
            if (count < 0) return false;
        }
        return count == 0;
    }




